Cannot list jobs with FilteredJob

Filtering jobs by time created throws `TypeError: 'tzinfo' is an invalid keyword argument for replace()`

I reproduced the error in a UT in this branch: https://github.com/microsoft/qdk-python/tree/guenp/fix-filtered-job

cc @anpaz @adelebai

```python
________________________________________________________________________________________________________________________ TestJob.test_filtered_job ________________________________________________________________________________________________________________________

self = <test_job.TestJob testMethod=test_filtered_job>

    def test_filtered_job(self):
        from datetime import date, timedelta
        from azure.quantum._client.models import JobStatus
    
        yesterday = date.today() - timedelta(days = 1)
        workspace = self.create_workspace()
>       jobs = workspace.list_jobs(created_after=yesterday)

tests/unit/test_job.py:134: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
azure/quantum/workspace.py:254: in list_jobs
    if deserialized_job.matches_filter(name_match, status, created_after):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <azure.quantum.job.Job object at 0x7f437f7f2fd0>, name_match = None, status = None, created_after = datetime.date(2021, 7, 15)

    def matches_filter(
        self,
        name_match: str = None,
        status:  Optional[JobStatus] = None,
        created_after: Optional[datetime] = None
    ) -> bool:
        """Checks if job (self) matches the given properties if any.
            :param name_match: regex expression for job name matching
            :param status: filter by job status
            :param created_after: filter jobs after time of job creation
        """
        if name_match is not None and re.search(name_match, self.details.name) is None:
           return False
    
        if status is not None and self.details.status != status.value:
            return False
    
>       if created_after is not None and self.details.creation_time.replace(tzinfo=timezone.utc) < created_after.replace(tzinfo=timezone.utc):
E       TypeError: 'tzinfo' is an invalid keyword argument for replace()

azure/quantum/job.py:122: TypeError
```

--- a/azure-quantum/azure/quantum/job.py
+++ b/azure-quantum/azure/quantum/job.py
@@ -2,7 +2,7 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 ##
-from datetime import datetime, timezone
+from datetime import date, datetime, timezone
 import logging
 import time
 import json
@@ -119,8 +119,17 @@ class Job:
         if status is not None and self.details.status != status.value:
             return False
         
-        if created_after is not None and self.details.creation_time.replace(tzinfo=timezone.utc) < created_after.replace(tzinfo=timezone.utc):
-            return False
+        if created_after is not None:
+            # if supplied date is date we must convert to datetime first
+            if type(created_after) is date:
+                created_after = datetime(created_after.year, created_after.month, created_after.day)
+            
+            # if supplied date is naive, assume local and convert to timezone aware object
+            if created_after.tzinfo is None:
+                created_after = created_after.astimezone()
+            
+            if self.details.creation_time.replace(tzinfo=timezone.utc) < created_after:
+                return False
 
         return True
 

--- a/azure-quantum/tests/unit/test_job.py
+++ b/azure-quantum/tests/unit/test_job.py
@@ -12,7 +12,7 @@ import time
 import os
 import functools
 import pytest
-from datetime import datetime, timedelta
+from datetime import date, datetime, timedelta
 
 from common import QuantumTestBase, ZERO_UID
 from azure.quantum import Job
@@ -160,6 +160,10 @@ class TestJob(QuantumTestBase):
 
             before_time = datetime.now() - timedelta(days=100)
             self.assertEqual(True, job.matches_filter(created_after=before_time))    
+
+            # test behaviour of datetime.date object
+            before_date = date.today() - timedelta(days=100)
+            self.assertEqual(True, job.matches_filter(created_after=before_date))
 
     def _test_job_submit(self, solver_type):
         """Tests the job submission and its lifecycle for a given solver.

